esmvalcore/_data_finder.py

```diff
@@ -33,29 +33,58 @@ def find_files(dirnames, filenames):
 
 
 def get_start_end_year(filename):
-    """Get the start and end year from a file name."""
+    """Get the start and end year from a file name.
+
+    Examples of allowed dates : 1980, 198001, 19801231,
+    1980123123, 19801231T23, 19801231T2359, 19801231T235959,
+    19801231T235959Z
+
+    Dates must be surrounded by - or _ or string start or string end
+    (after removing filename suffix)
+
+    Look first for two dates separated by - or _, then for one single
+    date, and if they are multiple, for one date at start or end
+    """
     stem = Path(filename).stem
     start_year = end_year = None
 
-    # First check for a block of two potential dates separated by _ or -
-    daterange = re.findall(r'([0-9]{4,12}[-_][0-9]{4,12})', stem)
+    date_pattern = r"(?P<year>[0-9]{4})(?P<month>[01][0-9]" + \
+                   "(?P<day>[0-3][0-9](T?(?P<hour>[0-2][0-9]" + \
+                   "(?P<minute>[0-5][0-9](?P<second>[0-5][0-9]Z?)?)?)?)?)?)?"
+    end_date_pattern = date_pattern.replace(">", "_end>")
+    date_range_pattern = date_pattern + r"[-_]" + end_date_pattern
+    #
+    # Next string allows to test that there is an allowed delimiter (or
+    # string start or end) close to date range (or to single date)
+    context = r"(?:^|[-_]|$)"
+    #
+    # First check for a block of two potential dates
+    date_range_pattern = context + date_range_pattern + context
+    daterange = re.search(date_range_pattern, stem)
     if daterange:
-        start_date, end_date = re.findall(r'([0-9]{4,12})', daterange[0])
-        start_year = start_date[:4]
-        end_year = end_date[:4]
+        start_year = daterange.group("year")
+        end_year = daterange.group("year_end")
     else:
         # Check for single dates in the filename
-        dates = re.findall(r'([0-9]{4,12})', stem)
+        single_date_pattern = context + date_pattern + context
+        dates = re.findall(single_date_pattern, stem)
         if len(dates) == 1:
-            start_year = end_year = dates[0][:4]
+            start_year = end_year = dates[0][0]
         elif len(dates) > 1:
             # Check for dates at start or end of filename
-            outerdates = re.findall(r'^[0-9]{4,12}|[0-9]{4,12}$', stem)
-            if len(outerdates) == 1:
-                start_year = end_year = outerdates[0][:4]
+            start = re.search(r'^' + date_pattern, stem)
+            end = re.search(date_pattern + r'$', stem)
+            if start:
+                if end is None:
+                    start_year = end_year = start.group('year')
+                else:
+                    pass  # NOk if two dates
```

Comment:
Can be removed. I don't understand the comment; can you make it clearer and place only the comment here?

Reply:
Actually , this construct is not correct :
```
if True :
        print("t")
else :
    # Com
    
print()
```
and one must pass, because finding two dates (one at start the other at end) is not considered correct (according to pre-existing code)

Reply:
What I meant is to leave out the entire else branch since
```python
if condition:
    do_something()
else:
    pass
```
is equivalent to
```python
if condition:
    do_something()
```
In any case, I think the comment needs to be more expressive. I do not understand what "NOk" is supposed to mean, nor the comment as a whole.

Reply:
OK. Will change accordingly